feat: GCP get-well [COMP-793] (#1316)
seqeralabs/docs: platform-enterprise_docs/compute-envs/google-cloud-batch.md

You can manage your key from the **Service Accounts** page.

**Workload Identity Federation**

:::tip

> Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication.

Reviewer: "for authentication, which are generated by Seqera Platform"

> Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication.

Reviewer: I think we need a callout here that unless you have generated and set an RSA keypair as part of your enterprise deployment set up that any authentication will fail with "**"WIF credentials require the OIDC provider to be configured (tower.oidc.pem.path)"** and link to the enterprise deployment with this information

> Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication.

Author: ```suggestion
Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication, which are generated by Seqera Platform.
```

:::

To authenticate using Workload Identity Federation, you must first configure a [Workload Identity Pool and Provider](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers) in your Google Cloud project. After setting up WIF in Google Cloud, you need the following information to create a credential in Seqera:

* **Service Account Email**: The email address of the Google Cloud service account that WIF will impersonate.
* **Workload Identity Provider**: The full resource path of the Workload Identity Provider (e.g., `projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/POOL_ID/providers/PROVIDER_ID`).

> * **Token Audience** (optional): The intended audience for the OIDC token. Configure this if your Workload Identity Provider requires a specific audience value.

Reviewer 2: @MichaelTansiniSeqera the comment is anchored to the **Token Audience** bullet but your proposed text is about user roles. I think the comment may have landed on the wrong line during review. Did you mean to comment on the **User permissions** list above (`roles/batch.jobsEditor`, `roles/iam.serviceAccountUser`, `roles/iam.serviceAccountViewer`)?

If that's the case, +1 on framing it as "minimum set of roles required", making that intent explicit in the doc would help admins justify the list to their security teams.

The Token Audience description as written is accurate. It's an optional field that's only needed when the WIF provider was created with a custom allowed audience; by default, Platform derives the audience from the provider path, so most setups can leave it blank. We could optionally tighten the wording to make the "leave blank unless..." default explicit, but no factual change needed.

Reviewer: np, I think we can resolve this

> WIF is supported across deployment environments including GKE (native support), AWS EKS, Azure AKS, and on-premises Kubernetes clusters with OIDC capabilities.

Reviewer: WIF here is a specifically Google authentication feature. It's supported as a credential for GCP Batch, Google Cloud and GKE

:::note
If you have not generated and set an RSA keypair as part of your Enterprise deployment, any authentication will fail with the message `WIF credentials require the OIDC provider to be configured (tower.oidc.pem.path)`.
:::

To authenticate using Workload Identity Federation, you must first configure a [Workload Identity Pool and Provider](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers) in your Google Cloud project. After setting up WIF in Google Cloud, you need the following information to create a credential in Seqera:

* **Service Account Email**: The email address of the Google Cloud service account that WIF will impersonate.
* **Workload Identity Provider**: The full resource path of the Workload Identity Provider (e.g., `projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/POOL_ID/providers/PROVIDER_ID`).
* **Token Audience** (optional): The intended audience for the OIDC token. Configure this if your Workload Identity Provider requires a specific audience value.

> WIF is supported across deployment environments including GKE (native support), AWS EKS, Azure AKS, and on-premises Kubernetes clusters with OIDC capabilities.

Author: ```suggestion
In the context of Seqera Platform, WIF is supported as a credential for GCP Batch, Google Cloud, and GKE.
```

You can manage your key from the **Service Accounts** page.

**Workload Identity Federation**

> Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication, which are generated by Seqera Platform.

Reviewer: Workflow Identity Federation also requires additional configuration of Seqera Platform when installing. For more details see below and [‎platform-enterprise_docs/enterprise/configuration/configtables/crypto_env.yml](https://github.com/seqeralabs/docs/pull/1316/changes#diff-2236463736954960b8a4ff8421e956b2f2655075be68981bba7e36ceba75c2cb) 

(or we could put this as an additional note?

**Generate the OIDC signing key**

Generate a PEM keypair and configure Platform to use it:

```bash
openssl genrsa -out private.pem 4096
openssl rsa -in private.pem -outform PEM -pubout -out public.pem
cat private.pem public.pem > oidc.pem
```

Set `TOWER_OIDC_PEM_PATH` to the path of the `oidc.pem` file in your Platform deployment. For example:

`TOWER_OIDC_PEM_PATH=/path/to/oidc.pem`.

If you have not generated and set an RSA keypair as part of your Enterprise deployment, any authentication will fail with the message `WIF credentials require the OIDC provider to be configured (tower.oidc.pem.path)`. 

After setting up WIF in Google Cloud, you need the following information to create a credential in Seqera:

* **Service Account Email**: The email address of the Google Cloud service account that WIF will impersonate.
* **Workload Identity Provider**: The full resource path of the Workload Identity Provider (e.g., `projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/POOL_ID/providers/PROVIDER_ID`).
* **Token Audience** (optional): The intended audience for the OIDC token. Configure this if your Workload Identity Provider requires a specific audience value.

:::caution

> If WIF authentication fails, verify that the Workload Identity Provider path is correctly formatted, the service account has the required permissions, and the Kubernetes service account is properly annotated for your deployment environment. Check the Seqera Platform logs for specific error details — a `400` error typically indicates an invalid provider format, while a `401` error indicates a token exchange failure.

Reviewer 3: ```suggestion
If WIF authentication fails, verify that the Workload Identity Provider path is correctly formatted, the service account has the required permissions, and the Kubernetes service account is properly annotated for your deployment environment. Check your Seqera Platform logs for specific error details. A `400` error typically indicates an invalid provider format, while a `401` error indicates a token exchange failure.
```

You can manage your key from the **Service Accounts** page.

**Workload Identity Federation**

Workload Identity Federation (WIF) is the recommended authentication method for production and regulated environments because it eliminates the need for long-lived service account keys. WIF uses short-lived OIDC tokens for authentication, which are generated by Seqera Platform. This requires the following steps in the GCP Console:

1. Create a [Workload Identity Pool and Provider](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers) in your Google Cloud project.

> 2. Set Seqera as an OIDC provider within the pool. Set the Issuer URL to https://{your-platform-domain}/api, where {your-platform-domain} is the hosted URL. The discovery endpoints are at /api/.well-known/openid-configuration and /api/.well-known/jwks.json, and both must be publicly reachable from GCP STS.

Reviewer 3: ```suggestion
2. Set Seqera as an OIDC provider within the pool. Set the Issuer URL to `https://{your-platform-domain}/api`, where `{your-platform-domain}` is the hosted URL. The discovery endpoints are at `/api/.well-known/openid-configuration` and `/api/.well-known/jwks.json`. Both must be publicly reachable from GCP STS.
```

3. Set the Allowed audiences. If left empty, GCP derives a default audience from the provider resource path in the format //iam.googleapis.com/projects/{PROJECT}/locations/global/workloadIden
  tityPools/{POOL}/providers/{PROVIDER}. If you specify a custom value, it must match exactly what you enter in the Token audience field when creating the Google WIF credential in Seqera.

> 4. Define an attribute mapping and condition. At a minimum set `google.subject=assertion.sub`. This maps the subject claim from Seqera's JWT to GCP's identity space. For more information see [here](https://docs.cloud.google.com/iam/docs/workload-identity-federation-with-other-providers#mappings-and-conditions)

Reviewer 3: ```suggestion
4. Define an attribute mapping and condition. At a minimum set `google.subject=assertion.sub`. This maps the subject claim from Seqera's JWT to GCP's identity space. For more information, see [Mappings and conditions](https://docs.cloud.google.com/iam/docs/workload-identity-federation-with-other-providers#mappings-and-conditions).
```

> 1. Create a [Workload Identity Pool and Provider](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers) in your Google Cloud project.
> 2. Set Seqera as an OIDC provider within the pool. Set the Issuer URL to https://{your-platform-domain}/api, where {your-platform-domain} is the hosted URL. The discovery endpoints are at /api/.well-known/openid-configuration and /api/.well-known/jwks.json, and both must be publicly reachable from GCP STS.
> 3. Set the Allowed audiences. If left empty, GCP derives a default audience from the provider resource path in the format //iam.googleapis.com/projects/{PROJECT}/locations/global/workloadIden
>   tityPools/{POOL}/providers/{PROVIDER}. If you specify a custom value, it must match exactly what you enter in the Token audience field when creating the Google WIF credential in Seqera.
> 4. Define an attribute mapping and condition. At a minimum set `google.subject=assertion.sub`. This maps the subject claim from Seqera's JWT to GCP's identity space. For more information see [here](https://docs.cloud.google.com/iam/docs/workload-identity-federation-with-other-providers#mappings-and-conditions)
> 5. Grant `roles/iam.workloadIdentityUser` on the service account created above to the Workload Identity Pool principal. This can be set for all pool identities or for a specific workspace.

Reviewer 3: ```suggestion
1. Create a [Workload Identity Pool and Provider](https://cloud.google.com/iam/docs/workload-identity-federation-with-other-providers) in your Google Cloud project.
2. Set Seqera as an OIDC provider within the pool. Set the Issuer URL to https://{your-platform-domain}/api, where {your-platform-domain} is the hosted URL. The discovery endpoints are at `/api/.well-known/openid-configuration` and `/api/.well-known/jwks.json`. Both must be publicly reachable from GCP STS.
3. Set the Allowed audiences. If left empty, GCP derives a default audience from the provider resource path in the format `//iam.googleapis.com/projects/{PROJECT}/locations/global/workloadIden
  tityPools/{POOL}/providers/{PROVIDER}`. If you specify a custom value, it must match exactly what you enter in the Token audience field when creating the Google WIF credential in Seqera.
4. Define an attribute mapping and condition. At a minimum set `google.subject=assertion.sub`. This maps the subject claim from Seqera's JWT to GCP's identity space. For more information see [here](https://docs.cloud.google.com/iam/docs/workload-identity-federation-with-other-providers#mappings-and-conditions)
5. Grant `roles/iam.workloadIdentityUser` on the service account created above to the Workload Identity Pool principal. This can be set for all pool identities or for a specific workspace.
```

```bash
openssl genrsa -out private.pem 4096
openssl rsa -in private.pem -outform PEM -pubout -out public.pem
cat private.pem public.pem > oidc.pem
```

Set `TOWER_OIDC_PEM_PATH` to the path of the `oidc.pem` file in your Platform deployment. For example:

`TOWER_OIDC_PEM_PATH=/path/to/oidc.pem`.

If you have not generated and set an RSA keypair as part of your Enterprise deployment, any authentication will fail with the message `WIF credentials require the OIDC provider to be configured (tower.oidc.pem.path)`.

After setting up WIF in Google Cloud, you need the following information to create a credential in Seqera:

* **Service Account Email**: The email address of the Google Cloud service account that WIF will impersonate.
* **Workload Identity Provider**: The full resource path of the Workload Identity Provider (e.g., `projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/POOL_ID/providers/PROVIDER_ID`).
* **Token Audience** (optional): The intended audience for the OIDC token. Configure this if your Workload Identity Provider requires a specific audience value. Ensure this matches what you have configured in the **Allowed Audiences** value in the GCP console.

> The issuer URL is ${TOWER_SERVER_URL}/api for all deployments, not just Cloud. In the GCP WIF provider, set the issuer to https://{your-platform-domain}/api regardless of whether it's Cloud or Enterprise. The discovery endpoints are at /api/.well-known/openid-configuration and /api/.well-known/jwks.json, and both must be publicly reachable from GCP STS.

Reviewer 3: ```suggestion
The issuer URL is `${TOWER_SERVER_URL}/api` for all deployments, not just Cloud. In the GCP WIF provider, set the issuer to `https://{your-platform-domain}/api` regardless of whether it's Cloud or Enterprise. The discovery endpoints are at `/api/.well-known/openid-configuration` and `/api/.well-known/jwks.json`, and both must be publicly reachable from GCP STS.
```

2. Enter a descriptive name for this environment, e.g., *Google Cloud Batch (europe-north1)*.
3. Select **Google Cloud Batch** as the target platform.

#### Credentials

1. From the **Credentials** drop-down, select existing Google credentials or select **+** to add new credentials. If you choose to use existing credentials, skip to the next section.
2. Enter a name for the credentials, e.g., *Google Cloud Credentials*.
3. Select the credential type:
   - **Google Service Account Key**: Paste the contents of the JSON key file created in the [service account keys](#authentication-methods) section.
   - **Google WIF**: Enter the **Service Account Email**, **Workload Identity Provider** path, and optionally the **Token Audience** as described in the [Workload Identity Federation](#authentication-methods) section.

#### Location and work directory

Select the **Location** where you will execute your pipelines. See [Location](https://cloud.google.com/compute/docs/regions-zones#available) to learn more.

In the **Pipeline work directory** field, enter your storage bucket URL, e.g., `gs://my-bucket`. This bucket must be accessible in the location selected in the previous step.

:::note

> When you specify a Cloud Storage bucket as your work directory, this bucket is used for the Nextflow [cloud cache](https://docs.seqera.io/nextflow/cache-and-resume#cache-stores) by default. You can specify an alternative cache location with the **Nextflow config file** field on the pipeline [launch](https://docs.seqera.io/platform-enterprise/launch/launchpad#launch-form) form.

Reviewer 3: ```suggestion
When you specify a Cloud Storage bucket as your work directory, this bucket is used for the Nextflow [cloud cache](https://docs.seqera.io/nextflow/cache-and-resume#cache-stores) by default. You can specify an alternative cache location with the **Nextflow config file** field on the pipeline [launch](../launch/launchpad#launch-form) form.
```

:::

#### Seqera features

Select **Enable Wave containers** to facilitate access to private container repositories and provision containers in your pipelines using the Wave containers service. See [Wave containers](https://docs.seqera.io/nextflow/wave) for more information.

> Select **Enable Fusion v2** to allow access to your Google Cloud Storage data via the [Fusion v2](https://docs.seqera.io/fusion) virtual distributed file system. This speeds up most data operations. The Fusion v2 file system requires Wave containers to be enabled. See [Fusion file system](https://docs.seqera.io/platform-enterprise/supported_software/fusion/overview) for configuration details.

Reviewer 3: ```suggestion
Select **Enable Fusion v2** to allow access to your Google Cloud Storage data via the [Fusion v2](https://docs.seqera.io/fusion) virtual distributed file system. This speeds up most data operations. The Fusion v2 file system requires Wave containers to be enabled. See [Fusion file system](../supported_software/fusion/overview) for configuration details.
```

4. Select the **Storage Admin** role, then select **Save**.

> :::tip
> You've created a project, enabled the necessary Google APIs, created a bucket, and created credentials for your service account. You now have what you need to set up a new compute environment in Seqera.
> :::

Reviewer 3: ```suggestion
```

:::

After your Google Cloud resources have been created, create a new Seqera compute environment:

1. In a workspace, select **Compute Environments > New Environment**.
2. Enter a descriptive name for this environment, e.g., *Google Cloud Batch (europe-north1)*.
3. Select **Google Cloud Batch** as the target platform.

#### Credentials

1. From the **Credentials** drop-down, select existing Google credentials or select **+** to add new credentials. If you choose to use existing credentials, skip to the next section.
2. Enter a name for the credentials, e.g., *Google Cloud Credentials*.
3. Select the credential type:
   - **Google Service Account Key**: Paste the contents of the JSON key file created in the [service account keys](#authentication-methods) section.
   - **Google WIF**: Enter the **Service Account Email**, **Workload Identity Provider** path, and optionally the **Token Audience** as described in the [Workload Identity Federation](#authentication-methods) section.

#### Location and work directory

Select the **Location** where you will execute your pipelines. See [Location](https://cloud.google.com/compute/docs/regions-zones#available) to learn more.

> In the **Pipeline work directory** field, enter your storage bucket URL, e.g., `gs://my-bucket`. This bucket must be accessible in the location selected in the previous step.

Reviewer 3: ```suggestion
In the **Pipeline work directory** field, enter your storage bucket URL. For example, `gs://my-bucket`. This bucket must be accessible in the location selected in the previous step.
```

#### Seqera features

Select **Enable Wave containers** to facilitate access to private container repositories and provision containers in your pipelines using the Wave containers service. See [Wave containers](https://docs.seqera.io/nextflow/wave) for more information.

Select **Enable Fusion v2** to allow access to your Google Cloud Storage data via the [Fusion v2](https://docs.seqera.io/fusion) virtual distributed file system. This speeds up most data operations. The Fusion v2 file system requires Wave containers to be enabled. See [Fusion file system](../supported_software/fusion/overview) for configuration details.

> Use Fusion v2

Reviewer 3: What's this for?

   :::note
   The **Instance Type** field sets a default machine type at the compute environment level. You can override this for individual processes using the `machineType` [process directive](https://docs.seqera.io/nextflow/google#process-definition) in your Nextflow configuration.
   :::

5. Use **Head Job CPUs** and **Head Job Memory** to specify the CPUs and memory allocated for the head job.
6. Use **Service Account email** to specify a service account email address other than the Compute Engine default to execute workflows with this compute environment (recommended for production environments).
7. Use **VPC** and **Subnet** to specify the name of a VPC network and subnet to be used by this compute environment. You can apply network tags directly in the **Network Tags** field (see below) or through VM instance templates used for the Nextflow head and compute jobs.

   :::note
   You must specify both a **VPC** and **Subnet** for your compute environment to use either.
   :::

> 8. Use **Network Tags** to apply GCP network tags to the compute instances in this environment. Network tags control which firewall rules and routing policies apply to your instances within their VPC.

Reviewer: "Network tags control which firewall rules and routing policies apply to your instances within a VPC"

Enable **Spot** to use Spot instances, which have significantly reduced cost compared to On-Demand instances.

:::note

> From Nextflow version 24.10, the default Spot reclamation retry setting changed to `0` on AWS and Google. By default, no internal retries are attempted on these platforms. Spot reclamations now lead to an immediate failure, exposed to Nextflow in the same way as other generic failures (returning for example, `exit code 1` on AWS). Nextflow will treat these failures like any other job failure unless you actively configure a retry strategy. For more information, see [Spot instance failures and retries](../troubleshooting_and_faqs/nextflow#spot-instance-failures-and-retries-in-nextflow).

Reviewer: Fusion snapshots is now possible on Google Batch. If that is selected I believe the reclamation strategy is set to 5 by default is it not @alberto-miranda ?

Reviewer 4: Yes, we set it by default to 5 minutes.

> From Nextflow version 24.10, the default Spot reclamation retry setting changed to `0` on AWS and Google. By default, no internal retries are attempted on these platforms. Spot reclamations now lead to an immediate failure, exposed to Nextflow in the same way as other generic failures (returning for example, `exit code 1` on AWS). Nextflow will treat these failures like any other job failure unless you actively configure a retry strategy. For more information, see [Spot instance failures and retries](https://docs.seqera.io/platform/24.2/troubleshooting_and_faqs/nextflow#spot-instance-failures-and-retries-in-nextflow).

Reviewer 3: This link is versioned?

> From Nextflow version 24.10, the default Spot reclamation retry setting changed to `0` on AWS and Google. By default, no internal retries are attempted on these platforms. Spot reclamations now lead to an immediate failure, exposed to Nextflow in the same way as other generic failures (returning for example, `exit code 1` on AWS). Nextflow will treat these failures like any other job failure unless you actively configure a retry strategy. For more information, see [Spot instance failures and retries](https://docs.seqera.io/platform/24.2/troubleshooting_and_faqs/nextflow#spot-instance-failures-and-retries-in-nextflow).

Author: ```suggestion
From Nextflow version 24.10, the default Spot reclamation retry setting changed to `0` on AWS and Google. By default, no internal retries are attempted on these platforms. Spot reclamations now lead to an immediate failure, exposed to Nextflow in the same way as other generic failures (returning for example, `exit code 1` on AWS). Nextflow will treat these failures like any other job failure unless you actively configure a retry strategy. For more information, see [Spot instance failures and retries](../troubleshooting_and_faqs/nextflow#spot-instance-failures-and-retries-in-nextflow).
```

:::

:::info
When a Spot instance is reclaimed by Google Cloud, Seqera Platform displays a human-readable description in the task details. Google Batch reserves exit codes in the 50001–59999 range for infrastructure events:

| Exit code | Description |
|-----------|-------------|
| 50001 | Spot instance was reclaimed by Google Cloud |
| 50002 | VM became unresponsive (host event or crash) |
| 50003 | VM unexpectedly rebooted during task execution |
| 50004 | Task reached unresponsive time limit and could not be cancelled |
| 50005 | Task exceeded maximum allowed runtime |

Exit codes 50006–59999 display a generic infrastructure failure message. Standard application exit codes (1–255) are displayed as before.
:::

> Apply [**Resource labels**](https://docs.seqera.io/platform-enterprise/resource-labels/overview) to the cloud resources consumed by this compute environment. Workspace default resource labels are prefilled.

Reviewer 3: ```suggestion
Apply [**Resource labels**](../resource-labels/overview) to the cloud resources consumed by this compute environment. Workspace default resource labels are prefilled.
```

#### Scripting and environment variables

* Expand **Staging options** to include:

>   + Optional [pre- or post-run Bash scripts](https://docs.seqera.io/platform-enterprise/launch/advanced#pre-and-post-run-scripts) that execute before or after the Nextflow pipeline execution in your environment.

Reviewer 3: ```suggestion
  + Optional [pre- or post-run Bash scripts](../advanced#pre-and-post-run-scripts) that execute before or after the Nextflow pipeline execution in your environment.
```

  + Global Nextflow configuration settings for all pipeline runs launched with this compute environment. Values defined here are pre-filled in the **Nextflow config file** field in the pipeline launch form. These values can be overridden during pipeline launch.

  :::info

>   Configuration settings in this field override the same values in the pipeline repository `nextflow.config` file. See [Nextflow config file](https://docs.seqera.io/platform-enterprise/launch/advanced#nextflow-config-file) for more information on configuration priority.

Reviewer 3: ```suggestion
  Configuration settings in this field override the same values in the pipeline repository `nextflow.config` file. See [Nextflow config file](../launch/advanced#nextflow-config-file) for more information on configuration priority.
```

  :::

* Specify custom **Environment variables** for the head and compute jobs.

#### Advanced options

:::note
If you use VM instance templates for the head or compute jobs (see step 8 below), resource allocation and networking values specified in the templates override any conflicting values you specify while creating your Seqera compute environment.
:::

1. Enable **Use Private Address** to ensure that your Google Cloud VMs aren't accessible to the public internet.
2. Use **Boot disk size** to control the persistent disk size that each task and the head job are provided.

> 3. Use **Boot Disk Image** to select a specific boot disk image for the compute instances. The dropdown is populated with available images from the GCP Compute API and supports autocomplete filtering. This field is optional — if not set, Google Batch uses the default image.

Reviewer 3: ```suggestion
3. Use **Boot Disk Image** to select a specific boot disk image for the compute instances. The dropdown is populated with available images from the GCP Compute API and supports autocomplete filtering. This field is optional. If not set, Google Batch uses the default image.
```

> 4. Use **Instance Type** to select a specific machine type for the compute instances. The dropdown is populated with available instance types for the selected region and supports autocomplete filtering. This field is optional — if not set, Google Batch selects an appropriate machine type automatically.

Reviewer 3: ```suggestion
4. Use **Instance Type** to select a specific machine type for the compute instances. The dropdown is populated with available instance types for the selected region and supports autocomplete filtering. This field is optional. If not set, Google Batch automatically selects an appropriate machine type.
```

   :::note
   The **Instance Type** field sets a default machine type at the compute environment level. You can override this for individual processes using the `machineType` [process directive](https://docs.seqera.io/nextflow/google#process-definition) in your Nextflow configuration.
   :::

5. Use **Head Job CPUs** and **Head Job Memory** to specify the CPUs and memory allocated for the head job.
6. Use **Service Account email** to specify a service account email address other than the Compute Engine default to execute workflows with this compute environment (recommended for production environments).
7. Use **VPC** and **Subnet** to specify the name of a VPC network and subnet to be used by this compute environment. You can apply network tags directly in the **Network Tags** field (see below) or through VM instance templates used for the Nextflow head and compute jobs.

   :::note
   You must specify both a **VPC** and **Subnet** for your compute environment to use either.
   :::

> 8. Use **Network Tags** to apply GCP network tags to the compute instances in this environment. Network tags control which firewall rules and routing policies apply to your instances within their VPC.

Author: ```suggestion
8. Use **Network Tags** to apply GCP network tags to the compute instances in this environment. Network tags control which firewall rules and routing policies apply to your instances within a VPC.
```
